File: scripts/generate-session-evaluation-sql.mjs
import fs from "node:fs";
import path from "node:path";

const root = process.cwd();
const sourcePath = path.join(root, "Mau_tin_nhan_nhan_xet_buoi_hoc_gui_phu_huynh.txt");
const baselinePath = path.join(root, "SQL session evaluations.sql");
const migrationPath = path.join(root, "SQL update session evaluation templates.sql");
const sourceLines = fs.readFileSync(sourcePath, "utf8").replace(/^\uFEFF/, "").split(/\r?\n/);

const statusDefinitions = [
  ["TIẾP THU KIẾN THỨC TỐT", "knowledge_good", "positive"],
  ["TẬP TRUNG VÀO BÀI GIẢNG", "focused", "positive"],
  ["SÔI NỔI", "enthusiastic", "positive"],
  ["TIẾP THU KIẾN THỨC CÒN CHẬM", "knowledge_slow", "needs_attention"],
  ["MẤT TẬP TRUNG", "distracted", "needs_attention"],
  ["ÍT TƯƠNG TÁC", "low_interaction", "needs_attention"],
  ["NÓI CHUYỆN RIÊNG", "private_talking", "needs_attention"],
  ["SỬ DỤNG ĐIỆN THOẠI", "phone_use", "needs_attention"],
  ["CHƯA LÀM BÀI TẬP", "homework_incomplete", "needs_attention"],
  ["ĐI HỌC MUỘN", "late", "needs_attention"],
  ["LỖI KHÁC", "other_behavior", "needs_attention"],
];
const cleanClosingHeading = "MẪU CÂU KẾT THÚC KHI KHÔNG CÓ LỖI";
const attentionClosingHeading = "MẪU CÂU KẾT THÚC KHI CÓ LỖI CẦN KHẮC PHỤC";
const headings = new Set([
  ...statusDefinitions.map(([heading]) => heading),
  cleanClosingHeading,
  attentionClosingHeading,
]);

function numberedItemsAfter(heading) {
  const start = sourceLines.findIndex(line => line.trim() === heading);
  if (start < 0) throw new Error(`Không tìm thấy mục "${heading}" trong file mẫu.`);
  const items = [];
  for (let index = start + 1; index < sourceLines.length; index += 1) {
    const line = sourceLines[index].trim();
    if (headings.has(line)) break;
    const match = line.match(/^\d+\.\s+(.+)$/);
    if (match) items.push(match[1].trim());
  }
  if (items.length !== 5) throw new Error(`Mục "${heading}" cần có đúng 5 mẫu, hiện có ${items.length}.`);
  return items;
}

function sql(value) {
  return `'${String(value).replaceAll("'", "''")}'`;
}

const statusRows = statusDefinitions.map(([name, code, category], index) =>
  `  (${sql(code)}, ${sql(name)}, ${sql(category)}, ${index + 1}, true)`
);
const templateDefinitions = [];
for (const [heading, code] of statusDefinitions) {
  for (const content of numberedItemsAfter(heading)) {
    templateDefinitions.push(["status", code, content]);
  }
}
for (const content of numberedItemsAfter(cleanClosingHeading)) {
  // The fixed opening is now assembled in JavaScript. The legacy "opening" slot stores clean closings.
  templateDefinitions.push(["opening", null, content]);
}
for (const content of numberedItemsAfter(attentionClosingHeading)) {
  templateDefinitions.push(["closing", null, content]);
}

const templateRows = templateDefinitions.map(([section, code, content]) => {
  const statusId = code
    ? `(SELECT id FROM public.evaluation_statuses WHERE code = ${sql(code)})`
    : "NULL";
  return `  (${sql(section)}, ${statusId}, ${sql(content)}, true, 1)`;
});
const dataSql = `-- Phiên bản mẫu nhận xét buổi học 2: khung tin nhắn cố định, cụm trạng thái và hai loại câu kết.
UPDATE public.evaluation_statuses SET active = false WHERE active;

INSERT INTO public.evaluation_statuses (code, name, category, display_order, active)
VALUES
${statusRows.join(",\n")}
ON CONFLICT (code) DO UPDATE SET
  name = EXCLUDED.name,
  category = EXCLUDED.category,
  display_order = EXCLUDED.display_order,
  active = EXCLUDED.active;

DELETE FROM public.evaluation_message_templates;

INSERT INTO public.evaluation_message_templates
  (section_type, status_id, content, active, weight)
VALUES
${templateRows.join(",\n")}
ON CONFLICT DO NOTHING;
`;

const migrationSql = `-- Cập nhật hệ thống nhận xét buổi học sang mẫu gửi phụ huynh phiên bản 2.
-- Có thể chạy lại an toàn; nhận xét đã gửi và bản nháp cũ không bị xóa.
BEGIN;

${dataSql}
COMMIT;

SELECT
  (SELECT count(*) FROM public.evaluation_statuses WHERE active) AS active_statuses,
  (SELECT count(*) FROM public.evaluation_message_templates WHERE active) AS active_templates;
`;

const baseline = fs.readFileSync(baselinePath, "utf8");
const dataMarker = "-- Phiên bản mẫu nhận xét buổi học 2:";
const markerStart = baseline.indexOf(dataMarker);
const dataStart = markerStart >= 0 ? markerStart : baseline.indexOf("INSERT INTO public.evaluation_statuses");
const summaryMatch = /SELECT\r?\n\s*\(SELECT count\(\*\) FROM public\.evaluation_statuses/.exec(baseline.slice(dataStart));
const summaryStart = summaryMatch ? dataStart + summaryMatch.index : -1;
if (dataStart < 0 || summaryStart < 0) throw new Error("Không xác định được vùng dữ liệu mẫu trong SQL nền.");
const updatedBaseline = baseline.slice(0, dataStart) + dataSql + "\n" + baseline.slice(summaryStart);

fs.writeFileSync(baselinePath, updatedBaseline, "utf8");
fs.writeFileSync(migrationPath, migrationSql, "utf8");
console.log(`Đã tạo ${path.basename(migrationPath)}: ${statusDefinitions.length} trạng thái, ${templateDefinitions.length} mẫu.`);

File: scripts/generate_session_evaluation_sql.py
from pathlib import Path
import json
import re

ROOT = Path(__file__).resolve().parent.parent
SOURCE = ROOT / "Mau_tin_nhan_nhan_xet_buoi_hoc_gui_phu_huynh.txt"
JS_GENERATOR = ROOT / "scripts" / "generate-session-evaluation-sql.mjs"
OUTPUT = ROOT / "SQL session evaluations.sql"

STATUSES = [
    ("RẤT TỐT", "very_good", "positive"),
    ("ỔN ĐỊNH", "stable", "neutral"),
    ("MẤT TẬP TRUNG", "distracted", "needs_attention"),
    ("ÍT TƯƠNG TÁC", "low_interaction", "needs_attention"),
    ("NÓI CHUYỆN RIÊNG", "private_talking", "needs_attention"),
    ("MẤT TRẬT TỰ", "disruptive", "needs_attention"),
    ("SỬ DỤNG ĐIỆN THOẠI", "phone_use", "needs_attention"),
    ("CHƯA LÀM BÀI TẬP THEO YÊU CẦU", "homework_incomplete", "needs_attention"),
    ("ĐI HỌC MUỘN", "late", "needs_attention"),
    ("HÀNH VI CHƯA PHÙ HỢP KHÁC", "other_behavior", "needs_attention"),
]


def numbered_lines(block):
    return [item.strip() for item in re.findall(r"^\s*\d+\.\s+(.+?)\s*$", block, re.M)]


def sql(value):
    return "'" + str(value).replace("'", "''") + "'"


source = SOURCE.read_text(encoding="utf-8-sig")
generator = JS_GENERATOR.read_text(encoding="utf-8")

opening = numbered_lines(source.split("## THƯ VIỆN MỞ ĐẦU", 1)[1].split("## THƯ VIỆN KẾT THÚC", 1)[0])
closing = numbered_lines(source.split("## THƯ VIỆN KẾT THÚC", 1)[1].split("# TRẠNG THÁI:", 1)[0])

private_block = re.search(
    r"const privateTalking = \{(.*?)\n\};",
    generator,
    re.S,
).group(1)
private_description_block, private_expectation_block = private_block.split("expectation:", 1)
private_description = [
    json.loads(item)
    for item in re.findall(r'"(?:[^"\\]|\\.)*"', private_description_block)
]
private_expectation = [
    json.loads(item)
    for item in re.findall(r'"(?:[^"\\]|\\.)*"', private_expectation_block)
]

status_blocks = {}
parts = re.split(r"^# TRẠNG THÁI:\s*", source, flags=re.M)
for part in parts[1:]:
    lines = part.splitlines()
    name = lines[0].strip()
    body = "\n".join(lines[1:])
    description, expectation = body.split("KỲ VỌNG", 1)
    status_blocks[name] = {
        "description": numbered_lines(description),
        "expectation": numbered_lines(expectation),
    }
status_blocks["NÓI CHUYỆN RIÊNG"] = {
    "description": private_description,
    "expectation": private_expectation,
}

if len(opening) != 10 or len(closing) != 10:
    raise RuntimeError("File cần có đúng 10 mẫu mở đầu và 10 mẫu kết thúc.")
for name, _, _ in STATUSES:
    item = status_blocks.get(name, {})
    if len(item.get("description", [])) != 10 or len(item.get("expectation", [])) != 10:
        raise RuntimeError(f'Trạng thái "{name}" chưa đủ 10 mẫu mô tả và 10 mẫu kỳ vọng.')

schema = re.search(
    r"const schema = String\.raw`(.*?)`;\n\nconst seed",
    generator,
    re.S,
).group(1).strip()

status_rows = [
    f"({sql(code)}, {sql(name)}, {sql(category)}, {index}, true)"
    for index, (name, code, category) in enumerate(STATUSES, 1)
]


def template_row(section, code, content):
    status_id = (
        f"(SELECT id FROM public.evaluation_statuses WHERE code = {sql(code)})"
        if code
        else "NULL"
    )
    return f"({sql(section)}, {status_id}, {sql(content)}, true, 1)"


template_rows = [
    *[template_row("opening", None, item) for item in opening],
    *[template_row("closing", None, item) for item in closing],
]
for name, code, _ in STATUSES:
    template_rows.extend(
        template_row("status", code, item)
        for item in status_blocks[name]["description"]
    )
    template_rows.extend(
        template_row("expectation", code, item)
        for item in status_blocks[name]["expectation"]
    )

status_values = ",\n  ".join(status_rows)
template_values = ",\n  ".join(template_rows)
seed = f"""
INSERT INTO public.evaluation_statuses (code, name, category, display_order, active)
VALUES
  {status_values}
ON CONFLICT (code) DO UPDATE SET
  name = EXCLUDED.name,
  category = EXCLUDED.category,
  display_order = EXCLUDED.display_order,
  active = EXCLUDED.active;

INSERT INTO public.evaluation_message_templates
  (section_type, status_id, content, active, weight)
VALUES
  {template_values}
ON CONFLICT DO NOTHING;

SELECT
  (SELECT count(*) FROM public.evaluation_statuses WHERE active) AS active_statuses,
  (SELECT count(*) FROM public.evaluation_message_templates WHERE active) AS active_templates;
""".strip()

OUTPUT.write_text(f"{schema}\n\n{seed}\n", encoding="utf-8")
print(f"Generated {OUTPUT.name}: {len(status_rows)} statuses, {len(template_rows)} templates.")
